tests/kernel/wave/wave_evoformer_test.py

```diff
@@ -0,0 +1,157 @@
+    ],
+)
+@pytest.mark.parametrize("dtype", [tkl.f16, tkl.bf16])
+def testEvoformerAttentionForward(
+    shape: tuple[int],
+    tile_sizes: tuple[int],
+    enable_scheduling: bool,
+    mfma_variant: MMAType,
+    dtype: DataType,
+    request,
+):
+    run_bench = request.config.getoption("--runperf")
+    dump_perf = request.config.getoption("--dump-perf-files-path")
+    shapes_and_tile_sizes = [(x, y) for x, y in zip(shape, tile_sizes)]
+    evoformer_fwd, symbols = get_evoformer_kernel(
+        *shapes_and_tile_sizes, mfma_variant, dtype
+    )
+
+    symbols.update(get_default_scheduling_params())
+
+    config = get_default_run_config()
+    if run_bench:
+        config["benchmark_batch_size"] = 1000
+        config["benchmark_repetitions"] = 3
+    if dump_perf is not None:
+        perf_filename = request.node.name + ".json"
+        config["benchmark_results_file"] = os.path.join(
+            dump_perf, "tk_" + perf_filename
+        )
+
+    with tk.gen.TestLaunchContext(
+        symbols,
+        canonicalize=True,
+        run=True,
+        run_bench=run_bench,
+        run_config=config,
+        schedule=enable_scheduling,
+        use_scheduling_barriers=enable_scheduling_barriers,
+    ):
+        torch.manual_seed(0)
+        if dtype == tkl.bf16:
+            torch_dtype = torch.bfloat16
+        else:
+            torch_dtype = torch.float16
+        batch, n, kv_seq_len, heads, head_dim, q_seq_len, v_dim = shape
+        q = device_randn(batch, n, q_seq_len, heads, head_dim, dtype=torch_dtype)
+        k = device_randn(batch, n, kv_seq_len, heads, head_dim, dtype=torch_dtype)
+        v = device_randn(batch, n, kv_seq_len, heads, v_dim, dtype=torch_dtype)
+        mask = device_randint(0, 2, (batch, n, kv_seq_len), dtype=torch_dtype)
+        mask_bias = 1e9 * (mask - 1)
+        bias = device_randn(batch, heads, q_seq_len, kv_seq_len, dtype=torch_dtype)
+        output = device_zeros(batch, n, q_seq_len, heads, v_dim, dtype=torch_dtype)
+        log2e = 1.44269504089
+        dk_sqrt = math.sqrt(1.0 / shape[4])
+        # TODO: Add scaling of QK as part of kernel.
+        # TODO: Add v-permute as part of kernel.
+        mb = evoformer_fwd(
+            q * dk_sqrt * log2e,
+            k,
+            v.permute([0, 1, 4, 3, 2]),
+            mask_bias,
+            bias * log2e,
+            output,
+        )
+
+        mask_bias = mask_bias.view([batch, n, 1, 1, kv_seq_len])
+        bias = bias.view([batch, 1, heads, q_seq_len, kv_seq_len])
+        torch_ref = attention_reference(q, k, v, [mask_bias, bias], dk_sqrt)
+
+        if test_dump_generated_mlir:
+            filename = f"wave_evoformer_{'x'.join(map(str, shape))}.mlir"
+            with open(filename, "w") as f:
+                f.write(mb.module_op.get_asm())
+
+        eps = 1e-2 if output.dtype == torch.float16 else 5e-2
+        print(f"Max diff: {torch.max(torch.abs(torch_ref - output)).item()}")
+        assert (
```

Comment:
Why not just `torch.allclose(..., atol=eps)`?

Reply:
This is a requirement from the customer.